src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AQNBuilder.Mono.cs

```diff
@@ -0,0 +1,483 @@
+
+				if (m_bUseAngleBracketsForGenerics)
+					Append('<');
+				else
+					Append('[');
+
+			}
+			public void CloseGenericArguments()
+			{
+				if (m_instNesting == 0)
+					Fail();
+				if (!CheckParseState(ParseState.START))
+					Fail();
+
+				m_parseState = ParseState.GENARGS;
+
+				m_instNesting --;
+
+				if (m_bFirstInstArg)
+				{
+					m_pStr.Remove(m_pStr.Length - 1, 1);
+				}
+				else
+				{
+					if (m_bUseAngleBracketsForGenerics)
+						Append('>');
+					else
+						Append(']');
+				}
+			}
+			public void OpenGenericArgument()
+			{
+				if(!CheckParseState (ParseState.START))
+					Fail ();
+
+				if (m_instNesting == 0)
+					Fail ();
+
+				m_parseState = ParseState.START;
+				m_bNestedName = false;
+
+				if (!m_bFirstInstArg)
+					Append(',');
+
+				m_bFirstInstArg = false;
+
+				if (m_bUseAngleBracketsForGenerics)
+					Append('<');
+				else
+					Append('[');
+				PushOpenGenericArgument();
+			}
+
+			public void CloseGenericArgument()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF | ParseState.ASSEMSPEC))
+					Fail();
+
+				if (m_instNesting == 0)
+					Fail();
+
+				m_parseState = ParseState.START;
+
+				if (m_bHasAssemblySpec)
+				{
+					if (m_bUseAngleBracketsForGenerics)
+						Append('>');
+					else
+						Append(']');
+				}
+
+				PopOpenGenericArgument();
+			}
+			public void AddName(string szName)
+			{
+				if (szName == null)
+					Fail ();
+
+				if (!CheckParseState(ParseState.START | ParseState.NAME))
+					Fail ();
+
+				m_parseState = ParseState.NAME;
+
+				if (m_bNestedName)
+					Append ('+');
+
+				m_bNestedName = true;
+
+				EscapeName (szName);
+			}
+
+			public void AddPointer()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail ();
+
+				m_parseState = ParseState.PTRARR;
+
+				Append ('*');
+			}
+
+			public void AddByRef()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.BYREF;
+
+				Append('&');
+			}
+
+			public void AddSzArray()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.PTRARR;
+
+				Append("[]");
+			}
+			public void AddArray(int rank)
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR))
+					Fail();
+
+				m_parseState = ParseState.PTRARR;
+
+				if (rank <= 0)
+					throw new ArgumentOutOfRangeException ();
+
+				if (rank == 1)
+					Append("[*]");
+				else if (rank > 64)
+				{
+					// Only taken in an error path, runtime will not load arrays of more than 32 dimentions
+					Append ($"[{rank}]");
+				}
+				else
+				{
+					var strDim = new StringBuilder (rank + 3);
+
+					strDim.Append ('[');
+					for(int i = 1; i < rank; i++) strDim.Append (',');
+					strDim.Append (']');
+					Append(strDim.ToString ());
```

Comment:
Why is a separate StringBuilder needed here?  Can't this just append to m_pStr directly?